diff --git a/ChangeLog b/ChangeLog
index 819c6fe..5e4d992 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,7 @@
+Wed Feb 19 17:47:01 2014  Tanaka Akira  <akr@fsij.org>
+
+	* ext/socket/ipsocket.c (ip_s_getaddress): Don't access freed memory.
+
 Sat Mar  1 21:00:27 2014  Marc-Andre Lafortune  <ruby-core@marc-andre.ca>
 
 	* proc.c: Having optional keyword arguments makes maximum arity +1,
diff --git a/ext/socket/ipsocket.c b/ext/socket/ipsocket.c
index 7b198bd..874c0da 100644
--- a/ext/socket/ipsocket.c
+++ b/ext/socket/ipsocket.c
@@ -305,12 +305,13 @@ ip_s_getaddress(VALUE obj, VALUE host)
 {
     union_sockaddr addr;
     struct addrinfo *res = rsock_addrinfo(host, Qnil, SOCK_STREAM, 0);
+    socklen_t len = res->ai_addrlen;
 
     /* just take the first one */
-    memcpy(&addr, res->ai_addr, res->ai_addrlen);
+    memcpy(&addr, res->ai_addr, len);
     freeaddrinfo(res);
 
-    return rsock_make_ipaddr(&addr.addr, res->ai_addrlen);
+    return rsock_make_ipaddr(&addr.addr, len);
 }
 
 void
